File: backend/app/schemas/knowledge.py
"""Knowledge-management schemas (Pydantic v2).

Grouped into:
- Legacy retrieval schemas (kept for the existing public search/list endpoints)
- Authoring schemas (create/update/detail)
- Lifecycle / review / feedback
- Taxonomy & ownership
- Versions, indexing, retrieval preview, analytics, citations
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

# ─────────────────────────────────────────────────────────────────────
# Legacy retrieval schemas (unchanged public contract)
# ─────────────────────────────────────────────────────────────────────


class KnowledgeArticleSchema(BaseModel):
    """Knowledge article response schema (legacy/public shape)."""

    id: str
    title: str
    category: str
    subcategory: str | None = None
    content: str
    steps: list[dict] = []
    tags: list[str] = []


class KnowledgeSearchRequest(BaseModel):
    """Search query for the knowledge base."""

    query: str = Field(..., min_length=3, max_length=1000)
    category: str | None = None
    limit: int = Field(default=5, ge=1, le=20)


class KnowledgeSearchResponse(BaseModel):
    """Search results from the knowledge base."""

    results: list[KnowledgeArticleSchema]
    total: int
    query: str


# ─────────────────────────────────────────────────────────────────────
# Structured authoring schemas
# ─────────────────────────────────────────────────────────────────────

ArticleStatus = Literal["draft", "in_review", "approved", "published", "archived"]
ArticleType = Literal["troubleshooting", "how_to", "faq", "known_error", "policy", "reference"]
Audience = Literal["employee", "it_staff", "admin"]
VisibilityScope = Literal["public_internal", "it_only", "admin_only"]


class StepSchema(BaseModel):
    """A single ordered troubleshooting/resolution step."""

    step_number: int = Field(..., ge=1)
    instruction: str = Field(..., min_length=1)
    details: str | None = None


class ArticleWriteBase(BaseModel):
    """Shared authoring fields for create/update."""

    title: str = Field(..., min_length=3, max_length=500)
    short_summary: str | None = Field(None, max_length=1000)
    article_type: ArticleType = "troubleshooting"
    language: str = "en"
    audience: Audience = "employee"
    visibility_scope: VisibilityScope = "public_internal"

    category: str = Field(..., min_length=1, max_length=100)
    subcategory: str | None = None
    product_or_system: str | None = None
    platform: str | None = None
    issue_type: str | None = None
    severity_hint: str | None = None
    tags: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    ownership_group_id: str | None = None

    content: str | None = None
    symptoms: list[str] = Field(default_factory=list)
    probable_causes: list[str] = Field(default_factory=list)
    prerequisites: list[str] = Field(default_factory=list)
    troubleshooting_steps: list[StepSchema] = Field(default_factory=list)
    resolution_steps: list[StepSchema] = Field(default_factory=list)
    validation_steps: list[StepSchema] = Field(default_factory=list)
    escalation_criteria: str | None = None
    escalation_target_team: str | None = None
    references: list[dict[str, Any]] = Field(default_factory=list)
    related_articles: list[str] = Field(default_factory=list)

    citation_label: str | None = None
    source_type: str | None = None
    source_reference: str | None = None
    review_interval_days: int | None = Field(default=180, ge=1, le=1095)


class ArticleCreate(ArticleWriteBase):
    """Payload to create a draft article."""


class ArticleUpdate(BaseModel):
    """Partial update payload — all fields optional."""

    title: str | None = Field(None, min_length=3, max_length=500)
    short_summary: str | None = None
    article_type: ArticleType | None = None
    language: str | None = None
    audience: Audience | None = None
    visibility_scope: VisibilityScope | None = None
    category: str | None = None
    subcategory: str | None = None
    product_or_system: str | None = None
    platform: str | None = None
    issue_type: str | None = None
    severity_hint: str | None = None
    tags: list[str] | None = None
    keywords: list[str] | None = None
    ownership_group_id: str | None = None
    content: str | None = None
    symptoms: list[str] | None = None
    probable_causes: list[str] | None = None
    prerequisites: list[str] | None = None
    troubleshooting_steps: list[StepSchema] | None = None
    resolution_steps: list[StepSchema] | None = None
    validation_steps: list[StepSchema] | None = None
    escalation_criteria: str | None = None
    escalation_target_team: str | None = None
    references: list[dict[str, Any]] | None = None
    related_articles: list[str] | None = None
    citation_label: str | None = None
    source_type: str | None = None
    source_reference: str | None = None
    change_summary: str | None = None


class ArticleSummary(BaseModel):
    """Lightweight article representation for list views."""

    id: str
    slug: str | None = None
    title: str
    short_summary: str | None = None
    status: ArticleStatus
    article_type: ArticleType
    category: str
    subcategory: str | None = None
    product_or_system: str | None = None
    platform: str | None = None
    audience: Audience
    version: int
    tags: list[str] = []
    ownership_group_id: str | None = None
    embedding_status: str
    quality_score: float | None = None
    view_count: int = 0
    usage_count: int = 0
    feedback_score: float | None = None
    next_review_due_at: str | None = None
    is_stale: bool = False
    updated_at: str | None = None


class ArticleDetail(ArticleSummary):
    """Full article representation for detail / editor views."""

    content: str | None = None
    keywords: list[str] = []
    issue_type: str | None = None
    severity_hint: str | None = None
    visibility_scope: VisibilityScope
    symptoms: list[str] = []
    probable_causes: list[str] = []
    prerequisites: list[str] = []
    troubleshooting_steps: list[dict[str, Any]] = []
    resolution_steps: list[dict[str, Any]] = []
    validation_steps: list[dict[str, Any]] = []
    escalation_criteria: str | None = None
    escalation_target_team: str | None = None
    references: list[dict[str, Any]] = []
    related_articles: list[str] = []
    citation_label: str | None = None
    source_type: str | None = None
    source_reference: str | None = None
    author_id: str | None = None
    reviewer_id: str | None = None
    approver_id: str | None = None
    confidence_level: float | None = None
    last_reviewed_at: str | None = None
    published_at: str | None = None
    archived_at: str | None = None
    indexed_at: str | None = None
    index_version: int = 0
    successful_resolution_count: int = 0
    negative_feedback_count: int = 0
    available_actions: list[str] = []
    created_at: str | None = None


class ArticleListResponse(BaseModel):
    articles: list[ArticleSummary]
    total: int
    limit: int
    offset: int


# ─────────────────────────────────────────────────────────────────────
# Lifecycle / review / feedback
# ─────────────────────────────────────────────────────────────────────


class LifecycleTransitionRequest(BaseModel):
    """Request to move an article through its lifecycle."""

    action: Literal[
        "submit_for_review",
        "approve",
        "request_changes",
        "reject",
        "publish",
        "archive",
        "restore",
        "create_revision",
    ]
    note: str | None = None
    change_summary: str | None = None


class ReviewNoteCreate(BaseModel):
    decision: Literal["comment", "approved", "rejected", "changes_requested"] = "comment"
    note: str = Field(..., min_length=1, max_length=5000)


class ReviewNoteSchema(BaseModel):
    id: str
    reviewer_id: str | None = None
    decision: str
    note: str | None = None
    from_status: str | None = None
    to_status: str | None = None
    created_at: str


class FeedbackCreate(BaseModel):
    rating: int | None = Field(None, ge=1, le=5)
    was_helpful: bool | None = None
    comment: str | None = Field(None, max_length=2000)
    source: Literal["portal", "chat", "ticket"] = "portal"
    resolved_issue: bool | None = None
    session_id: str | None = None


class FeedbackSchema(BaseModel):
    id: str
    article_id: str
    rating: int | None = None
    was_helpful: bool | None = None
    comment: str | None = None
    source: str
    created_at: str


# ─────────────────────────────────────────────────────────────────────
# Taxonomy & ownership
# ─────────────────────────────────────────────────────────────────────

TaxonomyType = Literal[
    "category", "subcategory", "product", "platform", "issue_type", "audience", "tag"
]


class TaxonomyTermCreate(BaseModel):
    term_type: TaxonomyType
    key: str = Field(..., min_length=1, max_length=120)
    label: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    parent_id: str | None = None
    ticket_category_mapping: str | None = None
    sort_order: int = 0


class TaxonomyTermUpdate(BaseModel):
    label: str | None = None
    description: str | None = None
    parent_id: str | None = None
    ticket_category_mapping: str | None = None
    sort_order: int | None = None
    is_active: bool | None = None


class TaxonomyTermSchema(BaseModel):
    id: str
    term_type: str
    key: str
    label: str
    description: str | None = None
    parent_id: str | None = None
    ticket_category_mapping: str | None = None
    sort_order: int = 0
    is_active: bool = True


class OwnershipGroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    display_name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    owner_id: str | None = None
    default_reviewer_id: str | None = None
    member_ids: list[str] = Field(default_factory=list)


class OwnershipGroupSchema(BaseModel):
    id: str
    name: str
    display_name: str
    description: str | None = None
    owner_id: str | None = None
    default_reviewer_id: str | None = None
    member_ids: list[str] = []
    is_active: bool = True


# ─────────────────────────────────────────────────────────────────────
# Versions, indexing, retrieval preview, analytics, citations
# ─────────────────────────────────────────────────────────────────────


class VersionSchema(BaseModel):
    id: str
    version: int
    title: str
    status: str
    change_summary: str | None = None
    author_id: str | None = None
    created_at: str


class VersionDetailSchema(VersionSchema):
    snapshot: dict[str, Any]


class ChunkSchema(BaseModel):
    chunk_index: int
    section: str
    header: str
    content: str
    token_estimate: int
    embedding_status: str


class RetrievalPreviewResponse(BaseModel):
    """How an article will be presented to the retrieval pipeline."""

    article_id: str
    citation_label: str
    chunking_strategy: str
    retrieval_text: str
    chunks: list[ChunkSchema]
    total_tokens: int
    warnings: list[str] = []


class ReindexRequest(BaseModel):
    article_ids: list[str] | None = None
    only_stale: bool = False


class IndexingStatusResponse(BaseModel):
    total_articles: int
    published_articles: int
    indexed_articles: int
    pending_articles: int
    stale_articles: int
    failed_articles: int
    total_chunks: int
    index_version: int
    last_indexed_at: str | None = None
    vector_store: str


class ReindexResult(BaseModel):
    requested: int
    reindexed: int
    chunks_written: int
    skipped: int
    errors: list[str] = []


class ArticleAnalyticsSchema(BaseModel):
    article_id: str
    title: str
    status: str
    view_count: int
    usage_count: int
    successful_resolution_count: int
    feedback_score: float | None = None
    negative_feedback_count: int
    resolution_rate: float | None = None
    quality_score: float | None = None
    is_stale: bool = False


class KnowledgeAnalyticsSummary(BaseModel):
    total_articles: int
    by_status: dict[str, int]
    published_articles: int
    stale_articles: int
    avg_quality_score: float | None = None
    total_views: int
    total_usage: int
    avg_resolution_rate: float | None = None
    top_articles: list[ArticleAnalyticsSchema]
    low_performers: list[ArticleAnalyticsSchema]


class CitationSchema(BaseModel):
    """A source attribution returned with grounded AI answers."""

    article_id: str
    title: str
    citation_label: str
    slug: str | None = None
    category: str | None = None
    snippet: str | None = None
    score: float | None = None


class RetrievalResultItem(BaseModel):
    article_id: str
    title: str
    category: str
    citation_label: str
    snippet: str
    score: float


class RetrievalResponse(BaseModel):
    results: list[RetrievalResultItem]
    citations: list[CitationSchema]
    confidence: float
    source: str
    published_only: bool
    low_confidence: bool


# ─────────────────────────────────────────────────────────────────────
# Quality, completeness, staleness, templates
# ─────────────────────────────────────────────────────────────────────


class DimensionScoreSchema(BaseModel):
    name: str
    label: str
    score: float
    earned: list[str] = []
    missing: list[str] = []


class CompletenessReportSchema(BaseModel):
    """Multi-dimension completeness report for the article editor."""

    score: float                    # 0 – 100
    grade: str                      # A / B / C / D / F
    dimensions: list[DimensionScoreSchema]
    ready_for_review: bool
    ready_for_publish: bool
    blocking_issues: list[str]      # hard blockers preventing publication
    suggestions: list[str]          # non-blocking improvements


class AuthorWarningSchema(BaseModel):
    """Inline warning item shown in the article editor."""

    severity: str           # "error" | "warning" | "info"
    field: str | None = None
    message: str
    guidance: str | None = None


class StaleAnalysisSchema(BaseModel):
    """Detailed staleness analysis for an article."""

    is_stale: bool
    staleness_score: float          # 0.0 (fresh) – 1.0 (very stale)
    days_since_update: int | None = None
    days_overdue: int | None = None
    reasons: list[str]
    recommendations: list[str]


class ArticleTemplateSchema(BaseModel):
    """A pre-filled article scaffold shown in the template picker."""

    key: str
    label: str
    category: str
    subcategory: str
    product_or_system: str
    description: str
    icon: str


class DuplicateHintSchema(BaseModel):
    """A lightweight hint returned when a similar article title is detected."""

    id: str
    title: str
    status: str

File: frontend/src/types/knowledge.ts
/** Knowledge management types — mirror of backend `schemas/knowledge.py`. */

export type ArticleStatus = 'draft' | 'in_review' | 'approved' | 'published' | 'archived';
export type ArticleType =
  | 'troubleshooting'
  | 'how_to'
  | 'faq'
  | 'known_error'
  | 'policy'
  | 'reference';
export type Audience = 'employee' | 'it_staff' | 'admin';
export type VisibilityScope = 'public_internal' | 'it_only' | 'admin_only';

export type LifecycleAction =
  | 'submit_for_review'
  | 'approve'
  | 'request_changes'
  | 'reject'
  | 'publish'
  | 'archive'
  | 'restore'
  | 'create_revision';

export interface Step {
  step_number: number;
  instruction: string;
  details?: string | null;
}

export interface ArticleSummary {
  id: string;
  slug?: string | null;
  title: string;
  short_summary?: string | null;
  status: ArticleStatus;
  article_type: ArticleType;
  category: string;
  subcategory?: string | null;
  product_or_system?: string | null;
  platform?: string | null;
  audience: Audience;
  version: number;
  tags: string[];
  ownership_group_id?: string | null;
  embedding_status: string;
  quality_score?: number | null;
  view_count: number;
  usage_count: number;
  feedback_score?: number | null;
  next_review_due_at?: string | null;
  is_stale: boolean;
  updated_at?: string | null;
}

export interface ArticleDetail extends ArticleSummary {
  content?: string | null;
  keywords: string[];
  issue_type?: string | null;
  severity_hint?: string | null;
  visibility_scope: VisibilityScope;
  symptoms: string[];
  probable_causes: string[];
  prerequisites: string[];
  troubleshooting_steps: Step[];
  resolution_steps: Step[];
  validation_steps: Step[];
  escalation_criteria?: string | null;
  escalation_target_team?: string | null;
  references: Array<Record<string, unknown>>;
  related_articles: string[];
  citation_label?: string | null;
  source_type?: string | null;
  source_reference?: string | null;
  author_id?: string | null;
  reviewer_id?: string | null;
  approver_id?: string | null;
  confidence_level?: number | null;
  last_reviewed_at?: string | null;
  published_at?: string | null;
  archived_at?: string | null;
  indexed_at?: string | null;
  index_version: number;
  successful_resolution_count: number;
  negative_feedback_count: number;
  available_actions: LifecycleAction[];
  created_at?: string | null;
}

export interface ArticleListResponse {
  articles: ArticleSummary[];
  total: number;
  limit: number;
  offset: number;
}

export interface ArticleWritePayload {
  title: string;
  short_summary?: string | null;
  article_type?: ArticleType;
  audience?: Audience;
  visibility_scope?: VisibilityScope;
  category: string;
  subcategory?: string | null;
  product_or_system?: string | null;
  platform?: string | null;
  issue_type?: string | null;
  severity_hint?: string | null;
  tags?: string[];
  keywords?: string[];
  ownership_group_id?: string | null;
  content?: string | null;
  symptoms?: string[];
  probable_causes?: string[];
  prerequisites?: string[];
  troubleshooting_steps?: Step[];
  resolution_steps?: Step[];
  validation_steps?: Step[];
  escalation_criteria?: string | null;
  escalation_target_team?: string | null;
  references?: Array<Record<string, unknown>>;
  related_articles?: string[];
  citation_label?: string | null;
  review_interval_days?: number;
}

export interface VersionSummary {
  id: string;
  version: number;
  title: string;
  status: string;
  change_summary?: string | null;
  author_id?: string | null;
  created_at: string;
}

export interface VersionDetail extends VersionSummary {
  snapshot: Record<string, unknown>;
}

export interface ReviewNote {
  id: string;
  reviewer_id?: string | null;
  decision: string;
  note?: string | null;
  from_status?: string | null;
  to_status?: string | null;
  created_at: string;
}

export interface FeedbackItem {
  id: string;
  article_id: string;
  rating?: number | null;
  was_helpful?: boolean | null;
  comment?: string | null;
  source: string;
  created_at: string;
}

export interface TaxonomyTerm {
  id: string;
  term_type: string;
  key: string;
  label: string;
  description?: string | null;
  parent_id?: string | null;
  ticket_category_mapping?: string | null;
  sort_order: number;
  is_active: boolean;
}

export interface OwnershipGroup {
  id: string;
  name: string;
  display_name: string;
  description?: string | null;
  owner_id?: string | null;
  default_reviewer_id?: string | null;
  member_ids: string[];
  is_active: boolean;
}

export interface RetrievalChunk {
  chunk_index: number;
  section: string;
  header: string;
  content: string;
  token_estimate: number;
  embedding_status: string;
}

export interface RetrievalPreview {
  article_id: string;
  citation_label: string;
  chunking_strategy: string;
  retrieval_text: string;
  chunks: RetrievalChunk[];
  total_tokens: number;
  warnings: string[];
}

export interface IndexingStatus {
  total_articles: number;
  published_articles: number;
  indexed_articles: number;
  pending_articles: number;
  stale_articles: number;
  failed_articles: number;
  total_chunks: number;
  index_version: number;
  last_indexed_at?: string | null;
  vector_store: string;
}

export interface ReindexResult {
  requested: number;
  reindexed: number;
  chunks_written: number;
  skipped: number;
  errors: string[];
}

export interface ArticleAnalytics {
  article_id: string;
  title: string;
  status: string;
  view_count: number;
  usage_count: number;
  successful_resolution_count: number;
  feedback_score?: number | null;
  negative_feedback_count: number;
  resolution_rate?: number | null;
  quality_score?: number | null;
  is_stale: boolean;
}

export interface KnowledgeAnalyticsSummary {
  total_articles: number;
  by_status: Record<string, number>;
  published_articles: number;
  stale_articles: number;
  avg_quality_score?: number | null;
  total_views: number;
  total_usage: number;
  avg_resolution_rate?: number | null;
  top_articles: ArticleAnalytics[];
  low_performers: ArticleAnalytics[];
}

export interface ArticleFilters {
  status?: ArticleStatus | '';
  category?: string;
  product_or_system?: string;
  platform?: string;
  audience?: Audience | '';
  ownership_group_id?: string;
  search?: string;
  review_due?: boolean;
  limit?: number;
  offset?: number;
}

// ─────────────────────────────────────────────────────────────────────
// Quality, completeness, stale analysis, templates
// ─────────────────────────────────────────────────────────────────────

export interface DimensionScore {
  name: string;
  label: string;
  score: number;            // 0.0 – 1.0
  earned: string[];
  missing: string[];
}

export interface CompletenessReport {
  score: number;            // 0 – 100
  grade: 'A' | 'B' | 'C' | 'D' | 'F';
  dimensions: DimensionScore[];
  ready_for_review: boolean;
  ready_for_publish: boolean;
  blocking_issues: string[];
  suggestions: string[];
}

export type WarningSeverity = 'error' | 'warning' | 'info';

export interface AuthorWarning {
  severity: WarningSeverity;
  field?: string | null;
  message: string;
  guidance?: string | null;
}

export interface StaleAnalysis {
  is_stale: boolean;
  staleness_score: number;         // 0.0 (fresh) – 1.0 (very stale)
  days_since_update?: number | null;
  days_overdue?: number | null;
  reasons: string[];
  recommendations: string[];
}

export interface ArticleTemplate {
  key: string;
  label: string;
  category: string;
  subcategory: string;
  product_or_system: string;
  description: string;
  icon: string;
}

export interface DuplicateHint {
  id: string;
  title: string;
  status: string;
}

